docs/resim/metrics/metrics_writer.md

```diff
@@ -0,0 +1,76 @@
+
+## How to write metrics
+
+The easiest way to write a new metric to the `ResimMetricsWriter` is to use the fluent API provided by the `Metric` class.
+
+Below is an example, making a `DoubleOverTimeMetric` using the fluent API - for more info on what this is, see [the Metrics Types docs](./metric_types.md).
+
+```python
+
+# Example metrics data - not necessarily written to metrics writer
+TIMESTAMPS: MetricsData = ...
+ERROR_INDEXED_BY_TIMESTAMPS: MetricsData = ...
+STATUSES_INDEXED_BY_TIMESTAMPS: MetricsData = ...
+
+# Example write
+metrics_writer
+      .add_double_over_time_metric("Localization error") # Type and name specified here
+      .with_description("Accumulated error in localization over time")
+      .append_doubles_over_time_data(ERROR_INDEXED_BY_TIMESTAMPS) # Append a single data series, as we only want to plot one
+      .append_statuses_over_time_data(STATUSES_INDEXED_BY_TIMESTAMPS) # Append associated statuses
+      .with_failure_definitions([DoubleFailureDefinition(fails_below=0.0, fails_above=1.0)])
+      .with_start_time(Timestamp(secs=0))
+      .with_end_time(Timestamp(secs=10))
+      .with_y_axis_name("Error (%)")
+      .with_legend_series_names(["Error"])
+      .with_status(MetricStatus.PASSED_METRIC_STATUS)
+      .with_importance(MetricImportance.HIGH_IMPORTANCE)
+      .with_should_display(True)
+      .with_blocking(False)
+```
+
+This will write a double over time metric to our output, with all the desired data and properties. As a rule of thumb, most of the parameters described in [the Metrics Types docs](./metric_types.md) can be set using the fluent API by prepending `with_` (or `append_`, for lists.)
+
+> NB: Notice that the `MetricsData` we used does *not* necessarily have to be written to the metrics writer. Provided the top-level metric referencing that data is written to the metrics writer, any associated data will also be written. For example, even the `TIMESTAMPS` data will (transitively) be written to our output file by the above code.
+
+### Writing metrics data
+
+You can also manually write `MetricsData` to the metrics_writer using 
+
+```python
+metrics_writer.add_metrics_data(TIMESTAMPS)
+```
+
+This is useful, because we may want to write data that is not immediately referenced by any metric. Such data may be used by **batch metrics**. The `MetricsData` class also has a simple fluent API to make this easier.
+
+## Validating the metrics writer output
+
+It's important to validate your output, to check it's a valid protobuf message that our system can plot. We currently provide this through the `validate_job_metrics` function in `resim.metrics.proto.validate_metrics_proto`. Note that this is called on the **output metrics message** of the metrics writer (i.e. `output.metrics_msg`), not the metrics writer itself.
```

Comment:
Do we do any automated validation on write from the metrics writer?

Reply:
Not as of right now - there are longer-form reasons for this, which in short are that the proto schema is still currently slightly unstable ground. For example, the statuses project has involved changing which statuses are permissible and which aren't. This means that the back-end needs to effectively support both of these versions, even though what is valid has changed. This is exacerbated by the fact we don't have good versioning set-up here. But I do think making sure we have a clear automatic validator should be a major priority for the metrics team.